README.md
## Submit more languages using this [form](https://goo.gl/forms/rNx4yAB9KM2fDXDG3)
#### Flatpak currently has limited functionality, can only send to Mastodon and no screenshots

# ShareXin

![ShareXin](https://sharexin.github.io/img/ShareX_Still.png "sharexin tweet")

[![GitHub Stars](https://img.shields.io/github/stars/thebitstick/ShareXin.svg?)](https://github.com/thebitstick/ShareXin)
[![Crates.io](https://img.shields.io/crates/v/sharexin.svg)](https://crates.io/crates/sharexin)

## Requirements
* Linux or FreeBSD
* rustc 1.30.1
* *scrot* (only need if `gnome-screenshot` not installed)
* *feh* (only need if `spectacle` is not installed or using GNOME Wayland)

## Features
* Takes screenshots
* Uploads to Twitter, Mastodon, and Imgur
* Saves screenshots to your Pictures
* Notifications
* GTK Dialog for entering a message with a tweet or toot
* Designed with **Wayland** in mind

### Desktop support
- GNOME desktop
- KDE Plasma desktop
- Budgie desktop
- Cinnamon desktop
- Unity desktop
- Any X11 DE

### Tested on
- Ubuntu 18.10
- Debian 9.5.0
- Fedora 29
- Arch Linux
- GhostBSD 18.10 (FreeBSD distro)

## Language support
* English
* Français (French) by [@Eleoryth](https://twitter.com/Eleoryth)
* Español (Spanish)
* Esperanto
* 简体中文 (Simplified Chinese)
* 繁體中文 (Traditional Chinese)
* 日本語 (Japanese)
* 한국어 (Korean)
* Deutsch (German) by [@qwertxzy](https://twitter.com/qwertxzy)
* Polski (Polish) by [@Michcioperz](https://twitter.com/Michcioperz)
* Português (Portuguese) by [@pillgp](https://twitter.com/pillgp)

## Compiling

#### Ubuntu 18.10 dependencies
* libgtk-3-dev
* libcairo2-dev
* libpango1.0-dev
* libgdk-pixbuf2.0-dev
* libatk1.0-dev
* libssl-dev
* libcurl4-openssl-dev
* libclang-dev
* build-essential

#### Debian 9.5.0 dependencies
* libgtk-3-dev
* libcairo2-dev
* libpango1.0-dev
* libgdk-pixbuf2.0-dev
* libatk1.0-dev
* libssl-dev
* libcurl4-openssl-dev
* libclang-dev
* build-essential

#### Arch Linux dependencies
* curl
* gtk3
* gdk-pixbuf2
* cairo
* glib2
* openssl
* dbus
* xcb-util
* base-devel
* clang

#### Fedora 29 dependencies
* gtk3-devel
* cairo-devel
* pango-devel
* gdk-pixbuf2-devel
* atk-devel
* openssl-devel
* libcurl-devel
* clang-devel

#### FreeBSD 11 dependencies
* openssl-devel
* gmake
* gcc
* dbus-glib
* devel/dbus
* gtk3
* devel/glib20 (via ports)
* cairo
* pango
* gdk-pixbuf2
* atk
* openssl
* curl

### Compiling
1. `git clone https://github.com/ShareXin/ShareXin/`
2. `cargo install`

**OR**
1. `cargo install sharexin`

## Changelog
#### [0.6.9] - 2018-11-16
- Removed Clipboard functionality (doesn't work under Wayland)
- Flatpak fixes (Ruby is still a pain, screenshotting doesn't work cause sandboxing doesn't allow access to host's gnome-screenshot, or spectacle)

#### [0.6.8] - 2018-11-02
- Moved anything in `cmd.rs` to `main.rs` (main only called cmd anyways)
- Moved anything in `save.rs` to `image.rs`
- `error.rs` is now `text.rs`, also merged with `language.rs`
- Reorganized things to accomodate for Flatpaks
- Added Flatpak!!!
- Tried to simplify a LOT of code
- `error.rs` now contains some planned enums for future rewritten error handling
- Fixed bug where some errors were not found in YAML files for some reason
- Fixed bug where because the `ShareXin` folder in pictures was already created, it attempted to display an error, which crashed for some reason
-
- Slightly modified Headerbar
- "Sent to Twitter" and "Sent to Mastodon" notifications disappear from your Notifications and don't stick around in the OSD

#### [0.6.7] - 2018-10-21
- Colors for character count on message popup match colors on Twitter Web
- Character count on message popup turns yellow when approaching limit with 20 characters left, just like on Twitter Web
- Didn't know `unreachable!("")` was a thing, replaced some instances of `panic!()` with it
- Removed useless `error::fatal()` message
- Disabling *"Ok"* button when no text is entered or when too much text is over service limit
- Removed "File saved" notification, not really needed
- Updated dependencies
- Debian is actually compatible **(had to regress from 3_22_30 of gtk-rs to 3_18)**
- Better comments
- `--upgrade` removed due to issues with openssl
- Character count only changes when keys are pressed in the TEXT Box, not the entire window
- Made `error.rs` actually readable
- Notifications for a sent image or status tweet/toot or Imgur post only last 2 seconds

#### [0.6.6] - 2018-10-31
- More testing done
- Removed some strange and unneccesary lines
- Removed lots of duplicate code
- Replaced many if else statements with match statements
- AppImage script provided
- General bug fixes
- Split screenshotting functionality off to [screenshot-rs](https://github.com/ShareXin/screenshot-rs)
- Removed swaywm support (even back when I used it, my implementation was trash, if any this is a service to sway users)
- Removed RefCell in `dialog.rs`
- Updated Twitter character limit (not full proof for non-Latin characters)
- Heavy rewrite of functions with clearer variables
- Old Twitter API restored for functionality, will be replaced by native API in a later update

#### [0.6.5] - 2018-09-26
- Work will continue on the project!
- Updated GNOME deps to 3_22_30 from 3_10
- Removed macOS support (how many people even used it?)
- Removed tray icon support (unneeded and unusable, GNOME doesn't even support it anymore)
- Updated various dependencies
- Removed ShareXin as a library
- Readded ShareXin to crates.io
- Added Subfolder under ShareXin in Pictures for sorting by years-month
- Rewrote method of determining screenshot selection software
  - Used to rely on hard-coded desktop variables to match with specific software
  - Now simply checks which software is installed, in a prefered order
- `t` and `toot` are still required for use with Twitter and Mastodon, but this WILL be addressed in 0.6.6
  - Native rust Twitter and Mastodon APIs will be added, with hopefully the same functionality.


Changelog.md
#### [0.6.4] - 2018-03-30
- Wow this project keeps going
- Added Portuguese translation

#### [0.6.3] - 2017-10-19
- Added support for Ubuntu 17.10 Gnome Desktop (congrats on upgrading to the future)

#### [0.6.2] - 2017-09-01
- System tray support (not for Mac)
- New slogan, guaranteed to work on Linux, FreeBSD, and macOS only
- Makefile and PKGBUILD
- Better command line parsing with clap-rs
- ShareXin library? I guess?

#### [0.6.1] - 2017-08-26
- XDG Directory for Pictures folder (switched my system to French, directories changed)

#### [0.6.0] - 2017-08-25
- macOS tested, Imgur works, `t` and `toot` should work if you have them
- Updated "upgrade" and error messages
- Figured out libcurl issue on some older systems
- Bug fixes

#### [0.5.9] - 2017-08-20
- ImageMagick integrated
- `killall vim` finally removed (kinda)!

#### [0.5.8] - 2017-08-20
- Imgur link now copied to clipboard _(should work on Wayland?)_
- Better desktop recognition
- Bug fixes

#### [0.5.7] - 2017-08-19
- Less repetitive code
- Area screenshots (except for KDE and Mac) use two separate files,
the frozen fullscreen and the part of it you select
- Fixed bug where Error 29 would always print (kind of a bug)

#### [0.5.6] - 2017-08-19
- Removes temporary file after sending it
- Budgie Desktop support
- Ubuntu Unity Desktop support
- Theoretical Mac support

#### [0.5.5] - 2017-08-18
- Added Polish translation
- Character count acts like ShareX, counts down from 140 or 500 (depending on destination)
- "File Saved" notification now includes image saved

#### [0.5.4] - 2017-08-17
- Just a Cargo fix, ignore

#### [0.5.3] - 2017-08-15
- Now packaging `t` (should work)
- Added `auth` option for Twitter and Mastodon

#### [0.5.2] - 2017-08-15
- GTk fixes
- Notification for error messages
- Bug fixes
- Now using AppImage for releases

#### [0.5.1] - 2017-08-14
- No more wildcards in getting modules
- Separated `language.rs`
- Now using scrot rather than maim
- `.kill()` exists for `Command`, no longer using killall feh
- Added Fatal error messages
- Cinnamon desktop support

#### [0.5.0] - 2017-08-13
- Rewrite of Language message system, now using templates, extremely easy to add more languages
- Bug fixes

#### [0.4.9] - 2017-08-12
- Lots of optimizations
- Actual bug fixes
- Better error messages
- `format!` is a thing
- `gnome-screenshot` has native shadow effects for windows
- Really dumb [seg](https://github.com/thebitstick/ShareXin/blob/db4b202d30eecb160b2e4db4fbd4f03f918ba4da/src/language.rs#L25) [fault](https://github.com/thebitstick/ShareXin/blob/db4b202d30eecb160b2e4db4fbd4f03f918ba4da/src/error.rs#L38)
- Less code, actually
- Less comments, cause "comments are bad"

#### [0.4.8] - 2017-08-11
- Stderr used for errors
- Returns for String methods rather than variables
- Completely remade the help function, universal syntax
- Bug fixes

#### [0.4.7] - 2017-08-09
- Character count now shows when you've hit the limit and when you've passed
the limit
- Rather than crashing, if notify-rust is unable to show you a notification,
it'll display an error message and exit nicely
- Added French error messages
- Bug fixes

#### [0.4.6] - 2017-08-09
- Error messages now translated
- When uploading an image to Twitter, the character limit is reduced to 117, just like on ShareX
- Less panicking
- Less repetitiveness
- t and toot now show notification if command fails to run (API troubles or no internet), rather than just showing you the "Sent" notification
- Bug fixes
- **Known issue**: In order to get t working correctly without a terminal, it must `killall vim`, so consider it a feature-bug

#### [0.4.5] - 2017-08-07
- BSD support, tested on FreeBSD with Xfce
- Rather than panicking, ShareXin exits with an error message
- Cleaner command line parsing
- Error messages now multi-lingual, that is if the error doesn't include $LANG
- Bug fixes

#### [0.4.4] - 2017-08-04
- Untested Mac notifications
- Shadows only added to Window screenshots
- Mac screenshot support coming soon
- Bug fixes? Cleaner code? Maybe

#### [0.4.3] - 2017-08-03
- Bug fixes
- Cleaner code

#### [0.4.1] - 2017-08-02
- Partial Wayland support for Gnome, Plasma, and Sway
- Gnome-Screenshot used on Gnome
- Spectacle used on Plasma/KDE
- Swaygrab used on Sway (i3-clone)

#### [0.4.0] - 2017-08-01
- Cursor hidden in all screenshots
- Removed double shadow
- Bug fixes

#### [0.3.9] - 2017-07-31
- Bug fixes

#### [0.3.8] - 2017-07-30
- Better struct management
- Imgur support! Opens image in browser
- Changed date variable (had to change it manually everytime

#### [0.3.7] - 2017-07-29
- Custom error handles
- Added experimental `open` Image Option, lets you select an image (animated or not) or possibly a video and send to a destination
- Update checker implemented properly
- Fixed service variables, now using structs
- Readying code for more destinations
- `auth` option not working at this current moment in time due to immature twitter apis available for Rust
- `imgur` option hints at what will be the next Destination

#### [0.3.6] - 2017-07-28
- Character count now turns red when over the limit
- You can't send a toot or tweet if it's over the limit
- Double the ImageMagick shadow
- No longer using Pipers crate
- Changed those weird "mort" and "morti" variables
- Commented most of the code, cleaned up

#### [0.3.5] - 2017-07-27
- Added button to check your image
- Added upgrade checker
- Added character count
- Beautified code
- New command line args

#### [0.3.4] - 2017-07-26
- Window screenshot adds shadow
- Separated main.rs function gui()

#### [0.3.3] - 2017-07-26
- New UI thanks to Glade
- Notifications now have language support
- Pressing Control+Return sends your message
- Native Buttons, language changes!

#### [0.3.2] - 2017-07-25
- Cleaned help messages, used GNU coreutil messages for some items

#### [0.3.1] - 2017-07-25
- Separated main.rs into {main, help, file, send}.rs, less scrolling

#### [0.3.0] - 2017-07-25
- Die Deutsche Sprache!

src/file.rs
// many _ are used in rust to make compiler ignore "unused_variable" tips

// args may be "-s" for selection screenshots
//     "-i" for window screenshots, or empty for fullscreenshots

#![allow(unused_variables)]

#[cfg(target_os = "macos")]
use screenshot::get_screenshot;

use std::time::Duration;
use std::process::*;
use std::*;
use std;
use notification;
use time;
use error;

pub fn open(file: String) {

    // tmp gets temporary dir
    let mut tmp = env::temp_dir();
    tmp.push("sharexin.png");

    // copy copies file to temp

    thread::sleep(Duration::new(1, 0));
    let _copy = match std::fs::copy(file, tmp.clone()) {
        Ok(ok) => ok,
        Err(e) => {
            println!("{}", error::message(13));
            process::exit(1)
        }
    };

    notification::file_saved();
}

#[cfg(target_os = "macos")]
fn screenshot(args: String, temp: &str) {

    // tmp gets temporary dir
    let mut tmp = env::temp_dir();
    tmp.push("sharexin.png");

    let screenshot = get_screenshot(0).unwrap();

    image::save_buffer(
        &Path::new(tmp),
        s.as_slice(),
        s.width() as u32,
        s.height() as u32,
        image::RGBA(8),
    ).unwrap();
}

#[cfg(not(target_os = "macos"))]
#[cfg(target_family = "unix")]
fn screenshot(args: String, temp: &str) {
    // x11/wayland session info gotten here
    let mut _session = String::new();
    _session = match env::var("XDG_SESSION_TYPE") {
        Ok(ok) => ok,
        Err(e) => {
            println!("{}", error::message(6));
            process::exit(1)
        }
    };
    let session = &_session.to_lowercase();

    // desktop environment info gotten here
    let mut _desktop = String::new();
    _desktop = match env::var("DESKTOP_SESSION") {
        Ok(ok) => ok,
        Err(e) => {
            println!("{}", error::message(7));
            process::exit(1)
        }
    };
    let desktop = &_desktop.to_lowercase();

    match session.as_ref() {
        "wayland" => match desktop.as_ref() {
            "gnome" => gnome(args, temp),
            "plasma" => kde(args, temp),
            "sway" => sway(args, temp),
            _ => {
                println!("{}", error::message(18));
                process::exit(1)
            }
        },
        "x11" => match desktop.as_ref() {
            "gnome" => gnome(args, temp),
            "plasma" => kde(args, temp),
            _ => maim(args, temp),
        },
        _ => {
            println!("{}", error::message(19));
            process::exit(1)
        }
    }
}

#[cfg(not(target_os = "macos"))]
#[cfg(target_family = "unix")]
fn sway(args: String, temp: &str) {
    if args == "-s" {

        // _before_image takes a full screenshot using swaygrab
        let _before_image = match Command::new("swaygrab").arg(temp.clone()).output() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(16));
                process::exit(1)
            }
        };

        // _feh displays it and sleeps the thread to wait for _image
        println!("Feh may not display properly due to tiling and Wayland.");

        let _feh = match Command::new("feh").arg(temp.clone()).arg("-F").spawn() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(21));
                process::exit(1)
            }
        };

        thread::sleep(Duration::new(0, 500000000));

        // _image lets use _slop to select
        let _slop = match Command::new("slop").output() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(23));
                process::exit(1)
            }
        };
        let slop = String::from_utf8_lossy(&_slop.stdout);
        let _image = match Command::new("convert")
            .args(&[temp.clone(), "-crop", &slop, temp.clone()])
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(22));
                process::exit(1)
            }
        };

        // _kill closes _feh, gently
        let _kill = Command::new("killall").arg("feh").output();

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else if args == "-i" {

        // _image uses swaygrab to get "focused" window and take screenshot
        let _image = match Command::new("swaygrab")
            .arg("-f")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(16));
                process::exit(1)
            }
        };

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else {

        // _image uses swaygrab to take screenshot
        let _image = match Command::new("swaygrab").arg(temp.clone()).status() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(16));
                process::exit(1)
            }
        };

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    }
}

#[cfg(not(target_os = "macos"))]
#[cfg(target_family = "unix")]
fn gnome(args: String, temp: &str) {
    if args == "-s" {

        // _before_image takes a full screenshot using gnome-screenshot
        let _before_image = match Command::new("gnome-screenshot")
            .arg("-f")
            .arg(temp.clone())
            .output()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(14));
                process::exit(1)
            }
        };

        // _feh displays it and sleeps the thread to wait for _image
        let _feh = match Command::new("feh").arg(temp.clone()).arg("-F").spawn() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(21));
                process::exit(1)
            }
        };

        thread::sleep(Duration::new(0, 500000000));

        // _image lets you select
        let _image = match Command::new("gnome-screenshot")
            .arg("-a")
            .args(&["-f", temp.clone()])
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(14));
                process::exit(1)
            }
        };

        // _kill closes _feh, gently
        let _kill = Command::new("killall").arg("feh").output();

        println!("gnome-screenshot doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else if args == "-i" {

        // _image uses gnome-screenshot to get current window and take screenshot
        let _image = match Command::new("gnome-screenshot")
            .arg("-w")
            .args(&["-f", temp.clone()])
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(14));
                process::exit(1)
            }
        };

        println!("gnome-screenshot doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else {

        // _image uses gnome-screenshot to take screenshot
        let _image = match Command::new("gnome-screenshot")
            .arg("-f")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(14));
                process::exit(1)
            }
        };

        println!("gnome-screenshot doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    }
}

#[cfg(not(target_os = "macos"))]
#[cfg(target_family = "unix")]
fn kde(args: String, temp: &str) {
    if args == "-s" {

        // _image pauses screen and lets you select
        let _image = match Command::new("spectacle")
            .arg("-rbno")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(15));
                process::exit(1)
            }
        };

        println!("spectacle doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else if args == "-i" {

        // _image uses spectacle to get current window and take screenshot
        let _image = match Command::new("spectacle")
            .arg("-abno")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(15));
                process::exit(1)
            }
        };

        println!("spectacle doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else {

        // _image uses spectacle to take screenshot
        let _image = match Command::new("spectacle")
            .arg("-fbno")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(15));
                process::exit(1)
            }
        };

        println!("spectacle doesnt give exit codes but maybe one day");
        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    }
}

#[cfg(not(target_os = "macos"))]
#[cfg(target_family = "unix")]
fn maim(args: String, temp: &str) {
    if args == "-s" {

        // _before_image takes a full screenshot using maim
        let _before_image = match Command::new("maim")
            .arg("--hidecursor")
            .arg(temp.clone())
            .output()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(17));
                process::exit(1)
            }
        };

        // _feh displays it and sleeps the thread to wait for _image
        let _feh = match Command::new("feh").arg(temp.clone()).arg("-F").spawn() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(21));
                process::exit(1)
            }
        };

        thread::sleep(Duration::new(0, 500000000));

        // _image lets you select
        let _image = match Command::new("maim")
            .args(&["--hidecursor", "-s"])
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(17));
                process::exit(1)
            }
        };

        // _kill closes _feh, gently
        let _kill = Command::new("killall").arg("feh").output();

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else if args == "-i" {

        // _xdo gets the active window
        let _xdo = match Command::new("xdotool").arg("getactivewindow").output() {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(20));
                process::exit(1)
            }
        };

        // _image uses maim to take the window gotten from xdo
        let xdo = String::from_utf8_lossy(&_xdo.stdout);

        let _image = match Command::new("maim")
            .args(&["--hidecursor", "-i"])
            .args(&[&xdo, temp.clone()])
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(17));
                process::exit(1)
            }
        };

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    } else {

        // _image uses maim to take screenshot
        let _image = match Command::new("maim")
            .arg("--hidecursor")
            .arg(temp.clone())
            .status()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(17));
                process::exit(1)
            }
        };

        if _image.code() == Some(1) {
            println!("Exiting...");
            process::exit(1);
        }
    }
}

pub fn image(args: String) {
    //  tmp gets the temporary directory of the system
    let mut tmp = env::temp_dir();

    // adds filename
    tmp.push("sharexin.png");

    // makes a string
    let temp = tmp.to_str().unwrap();

    screenshot(args.clone(), temp.clone());
    if !tmp.exists() {
        println!("{}", error::message(13));
        process::exit(1);
    }

    if args == "-i" {
        //  adds a shadow
        let _ = match Command::new("convert")
            .arg(temp.clone())
            .args(&[
                "(",
                "+clone",
                "-background",
                "black",
                "-shadow",
                "80x3+5+5",
                ")",
                "+swap",
                "-background",
                "none",
                "-layers",
                "merge",
                "+repage",
            ])
            .arg(temp.clone())
            .spawn()
        {
            Ok(ok) => ok,
            Err(e) => {
                println!("{}", error::message(22));
                process::exit(1)
            }
        };
    }

    save();
}

pub fn save() {
    // tmp gets temporary dir
    let mut tmp = env::temp_dir();
    tmp.push("sharexin.png");

    // home gets the user's name
    let home = match env::var("HOME") {
        Ok(home) => home,
        Err(e) => {
            println!("{}", error::message(0));
            process::exit(1)
        }
    };
    let mut pictures = String::from(home.clone());
    pictures.push_str("/Pictures/ShareXin");

    // _dir creates pictures dir if not already there
    let _dir = match std::fs::create_dir(pictures) {
        Ok(ok) => ok,
        Err(e) => println!("{}", error::message(25)),
    };
    let mut new_file = String::from(home);
    new_file.push_str("/Pictures/ShareXin/sharexin-");

    // time gets the time in a nice format
    let time = String::from(match time::strftime("%Y-%m-%d-%H_%M_%S", &time::now()) {
        Ok(ok) => ok,
        Err(e) => {
            println!("{}", error::message(12));
            process::exit(1)
        }
    });
    new_file.push_str(&time);
    new_file.push_str(".png");

    // _clone copies the temp file to your home pic dir
    thread::sleep(Duration::new(1, 0));
    let _clone = match std::fs::copy(tmp.clone(), new_file) {
        Ok(ok) => ok,
        Err(e) => {
            println!("{}", error::message(13));
            process::exit(1)
        }
    };
    notification::file_saved();
}
